### PR TITLE
add kubectl tests

### PR DESCRIPTION
**Pull request by sed-i**
This PR adds kubectl itests for spotting the various edge cases @mmanciop is repeatedly finding manually.

### CODE REVIEW

File: tests/integration/test_kubectl_delete.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        )
+        await ops_test.model.wait_for_idle(apps=[app_name], status="active", timeout=1000)
+        await check_prometheus_is_ready(ops_test, app_name, 0)
+
+
+@pytest.mark.abort_on_fail
+async def test_kubectl_delete_pod(ops_test, prometheus_charm):
+    pod_name = f"{app_name}-0"
+    cmd = [
+        "sudo",
```
Comment (sed-i) [inline tests/integration/test_kubectl_delete.py:43]: I have sudo here for now because I wasn't able to correctly escape quotation marks for `sg -c`:
```python
cmd = [                                                                                        
    "sg",                                                                                      
    "-c",                                                                                      
    "microk8s",                                                                                
    r'"{}"'.format(                                                                            
        " ".join(                                                                              
            [                                                                                  
                "microk8s.kubectl",                                                            
                "delete",                                                                      
                "pod",                                                                         
                "-n",                                                                          
                ops_test.model_name,                                                           
                pod_name,                                                                      
            ]                                                                                  
        )                                                                                      
    ),                                                                                         
] 
```

**Comment by mmanciop** [comment]
Cool stuff! But I object the implications pf defining these “edge cases”. Upgrades are a main concern for an operator.

**Review by mmanciop — COMMENTED** [review]
The logic LGTM, but I have not made due diligence for the functional verification

File: tests/integration/test_kubectl_delete.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+METADATA = yaml.safe_load(Path("./metadata.yaml").read_text())
+app_name = METADATA["name"]
+
+
+@pytest.mark.abort_on_fail
+async def test_deploy_from_local_path(ops_test, prometheus_charm):
+    """Deploy the charm-under-test."""
+    resources = {"prometheus-image": METADATA["resources"]["prometheus-image"]["upstream-source"]}
+    log.info("deploy local charm")
```
Comment (simskij) [inline tests/integration/test_kubectl_delete.py:26]:
```suggestion
    log.debug("deploy local charm")
```

let's try to keep the test logs clean 👍🏽

File: tests/integration/test_kubectl_delete.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+    async with IPAddressWorkaround(ops_test):
+        await ops_test.model.deploy(
+            prometheus_charm, application_name=app_name, resources=resources
+        )
+        await ops_test.model.block_until(
+            lambda: len(ops_test.model.applications[app_name].units) > 0
+        )
+        await ops_test.model.wait_for_idle(apps=[app_name], status="active", timeout=1000)
```
Comment (simskij) [inline tests/integration/test_kubectl_delete.py:35]: good candidate for gathering. i also have a vague recollection of reading somewhere that `wait_for_idle` now also blocks until there is at least one unit that is idle?

File: tests/integration/test_kubectl_delete.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        ops_test.model_name,
+        pod_name,
+    ]
+    log.info("Removing pod '%s' from model '%s' with cmd: %s", pod_name, ops_test.model_name, cmd)
+
+    retcode, stdout, stderr = await ops_test._run(*cmd)
+    assert retcode == 0, f"kubectl failed: {(stderr or stdout).strip()}"
+    log.info(stdout)
+    await ops_test.model.block_until(lambda: len(ops_test.model.applications[app_name].units) > 0)
+    await ops_test.model.wait_for_idle(apps=[app_name], status="active", timeout=1000)
```
Comment (simskij) [inline tests/integration/test_kubectl_delete.py:57]: same comment as above.

File: tests/integration/test_kubectl_delete.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        "pod",
+        "-n",
+        ops_test.model_name,
+        pod_name,
+    ]
+    log.info("Removing pod '%s' from model '%s' with cmd: %s", pod_name, ops_test.model_name, cmd)
+
+    retcode, stdout, stderr = await ops_test._run(*cmd)
+    assert retcode == 0, f"kubectl failed: {(stderr or stdout).strip()}"
+    log.info(stdout)
```
Comment (simskij) [inline tests/integration/test_kubectl_delete.py:55]:
```suggestion
    log.debug(stdout)
```

File: tests/integration/test_kubectl_delete.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+log = logging.getLogger(__name__)
+
+METADATA = yaml.safe_load(Path("./metadata.yaml").read_text())
+app_name = METADATA["name"]
+
+
+@pytest.mark.abort_on_fail
+async def test_deploy_from_local_path(ops_test, prometheus_charm):
+    """Deploy the charm-under-test."""
+    resources = {"prometheus-image": METADATA["resources"]["prometheus-image"]["upstream-source"]}
```
Comment (simskij) [inline tests/integration/test_kubectl_delete.py:25]: `upstream_source` could be extracted after `app_name` and then just used here to avoid cluttering the function body.

File: tests/integration/test_kubectl_delete.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    cmd = [
+        "sudo",
+        "microk8s.kubectl",
+        "delete",
+        "pod",
+        "-n",
+        ops_test.model_name,
+        pod_name,
+    ]
+    log.info("Removing pod '%s' from model '%s' with cmd: %s", pod_name, ops_test.model_name, cmd)
```
Comment (simskij) [inline tests/integration/test_kubectl_delete.py:51]:
```suggestion
    log.debug("Removing pod '%s' from model '%s' with cmd: %s", pod_name, ops_test.model_name, cmd)
```

**Review by simskij — CHANGES_REQUESTED** [review]

File: tests/integration/test_kubectl_delete.py (hunk 6 above)
Comment (sed-i) [reply tests/integration/test_kubectl_delete.py]: Moved to global scope

File: tests/integration/test_kubectl_delete.py (hunk 2 above)
Comment (sed-i) [reply tests/integration/test_kubectl_delete.py]: Done

File: tests/integration/test_kubectl_delete.py (hunk 7 above)
Comment (sed-i) [reply tests/integration/test_kubectl_delete.py]: Done

File: tests/integration/test_kubectl_delete.py (hunk 5 above)
Comment (sed-i) [reply tests/integration/test_kubectl_delete.py:52]: I think the output of `_run` qualifies for `info`. What's the criterion?

File: tests/integration/test_kubectl_delete.py (hunk 3 above)
Comment (sed-i) [reply tests/integration/test_kubectl_delete.py]: Correct. Removed `block_until`.

File: tests/integration/test_kubectl_delete.py (hunk 5 above)
Comment (simskij) [reply tests/integration/test_kubectl_delete.py:52]: i don't think we have clear criteria. personally, i'd say: logs should be completely empty unless someone turns on debugging or something goes wrong.

Comment (sed-i) [reply tests/integration/test_kubectl_delete.py]: Done

File: tests/integration/test_kubectl_delete.py (hunk 1 above)
Comment (rbarry82) [reply tests/integration/test_kubectl_delete.py:40]:
```
cmd = [
  "sg",
  "microk8s",
  "-c",
  " ".join(
    [
      "microk8s.kubectl",
      "delete",
      "pod",
      "-n",
      ops_test.model_name,
      pod_name
    ]
  )
]

File: tests/integration/test_kubectl_delete.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+async def test_deploy_from_local_path(ops_test, prometheus_charm):
+    """Deploy the charm-under-test."""
+    logger.debug("deploy local charm")
+
+    async with IPAddressWorkaround(ops_test):
+        await ops_test.model.deploy(
+            prometheus_charm, application_name=app_name, resources=resources
+        )
+        await ops_test.model.wait_for_idle(apps=[app_name], status="active", timeout=1000)
+        await check_prometheus_is_ready(ops_test, app_name, 0)
```
Comment (Abuelodelanada) [inline tests/integration/test_kubectl_delete.py:33]: We are avoiding inner functions ;-)

**Review by Abuelodelanada — APPROVED** [review]

File: tests/integration/test_kubectl_delete.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+        "microk8s",
+        "-c",
+        " ".join(["microk8s.kubectl", "delete", "pod", "-n", ops_test.model_name, pod_name]),
+    ]
+
+    logger.debug(
+        "Removing pod '%s' from model '%s' with cmd: %s", pod_name, ops_test.model_name, cmd
+    )
+
+    retcode, stdout, stderr = await ops_test._run(*cmd)
```
Comment (dstathis) [inline tests/integration/test_kubectl_delete.py:51]: Not a blocker by any means but I think you should not use a private method like this. subprocess will work just as well and will never change.

Comment (rbarry82) [reply tests/integration/test_kubectl_delete.py:51]: It's also [available as `.run()`](https://github.com/charmed-kubernetes/pytest-operator/blob/f1dccecaaa05162b02173c16e841cf57aafd3928/pytest_operator/plugin.py#L222). It does inject environment variables for Juju, but we probably don't really need those, and `asyncio.create_subprocess_exec` may be enough on its own considering we're not actually interacting with Juju in this test.

**Review by dstathis — APPROVED** [review]
This comment is out of scope for this PR, but, I want to mention it anyway. `IPAddressWorkaround` should really be called something like `set_update_status_interval` as it has nothing to do with IP addresses.

File: tests/integration/test_kubectl_delete.py (hunk 9 above)
Comment (sed-i) [reply tests/integration/test_kubectl_delete.py:51]: Thanks for pointing this out. This is a copy paste from old "idiomatic" pytest, and now
`_run = run  # backward compatibility alias`
I will change this to `run`.

File: tests/integration/test_kubectl_delete.py (hunk 8 above)
Comment (Abuelodelanada) [reply tests/integration/test_kubectl_delete.py:33]: I do not know why I saw a `def` here

**Review by Abuelodelanada — APPROVED** [review]

**Review by dstathis — APPROVED** [review]
